packages/hash/api/src/graphql/resolvers/knowledge/page/update-page-contents.ts

```diff
@@ -0,0 +1,443 @@
+import { ApolloError, UserInputError } from "apollo-server-errors";
+import produce from "immer";
+import { GraphApi } from "@hashintel/hash-graph-client";
+
+import {
+  BlockModel,
+  EntityModel,
+  PageModel,
+  UserModel,
+} from "../../../../model";
+import { exactlyOne } from "../../../../util";
+import {
+  KnowledgeCreateEntityAction,
+  KnowledgeEntityDefinition,
+  KnowledgeInsertBlockAction,
+  KnowledgeSwapBlockDataAction,
+  KnowledgeUpdateEntityAction,
+  KnowledgeUpdatePageAction,
+  KnowledgeUpdatePageContentsResult,
+  MutationKnowledgeUpdatePageContentsArgs,
+  ResolverFn,
+} from "../../../apiTypes.gen";
+import { LoggedInGraphQLContext } from "../../../context";
+import { pageModelToGQL, UnresolvedPageGQL } from "../model-mapping";
+
+type UpdatePageActionKey = keyof KnowledgeUpdatePageAction;
+
+const filterForAction = <T extends UpdatePageActionKey>(
+  actions: KnowledgeUpdatePageAction[],
+  key: T,
+): { action: NonNullable<KnowledgeUpdatePageAction[T]>; index: number }[] =>
+  actions.reduce<
+    { action: NonNullable<KnowledgeUpdatePageAction[T]>; index: number }[]
+  >((acc, current, index) => {
+    if (current != null && key in current) {
+      acc.push({ action: current[key]!, index });
+    }
+    return acc;
+  }, []);
```

Comment:
I did consider doing that, and I agree that it would probably be 'best', but it does seem like the actions list is fairly small most of the time, so traversal is very quick. I can add a todo, though :)

Reply:
I didn't add a todo, but a note on optimization b53fa0d31eae444d38b39ee1261d010ee0d5be50